### PR TITLE
Support codellama

### PR DESCRIPTION
**Pull request by lvhan028**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

support codellama

## Modification
- [x] chat.py: code input
- [x] client.py: code input
- [x] model.py: chat template
- [x] deploy.py: rope_theta
- [x] turbomind csrc: rotary_emb_base from config.ini
- [x] docs

### compatibility
- [x] Use lmdeploy v0.0.7 to convert the following models' layout and use turbomind of this PR to do inference
    - [x]  llama2-7b-chat (FP16/int4)
    - [x]  llama2-70b-chat (FP16, tp)
    - [x] internlm-chat-7b (FP16)
    - [x] qwen-7b-chat (FP16)
    - [x] internal puyu
### functionality
- [x] convert model and inference it by this PR
    - [x] codellama-7b-chat (FP16)
    - [x] codellama-34b-chat (FP16, tp)
    - [x]  llama2-7b-chat (FP16/int4)
    - [x]  llama2-70b-chat (FP16, tp)
    - [x] internlm-chat-7b (FP16)
    - [x] qwen-7b-chat (FP16)
    - [x] baichuan2-7b-chat(FP16)
    - [x] internal puyu

### CODE REVIEW

File: src/turbomind/kernels/decoder_masked_multihead_attention.h [hunk 1]
```diff
@@ -121,6 +121,7 @@ struct Multihead_attention_params: public Multihead_attention_params_base<T> {
     int        max_position_embeddings    = 0;
     bool       use_dynamic_ntk            = false;
     bool       use_logn_attn              = false;
+    float      rotary_emb_base            = 10000.0f;
```
Comment (lzhangzz) [inline src/turbomind/kernels/decoder_masked_multihead_attention.h:124]: rotary_embedding_base

File: src/turbomind/kernels/unfused_attention_kernels.cu [hunk 2]
```diff
@@ -863,6 +863,7 @@ __global__ void add_fusedQKV_bias_transpose_kernel(T* q_buf,
                                                    int        kv_head_num,
                                                    int        size_per_head,
                                                    int        rotary_embedding_dim,
+                                                   float      rotary_emb_base,
```
Comment (lzhangzz) [inline src/turbomind/kernels/unfused_attention_kernels.cu:866]: rotary_emebdding_base

File: src/turbomind/models/llama/llama_params.h [hunk 3]
```diff
@@ -5,10 +5,11 @@
 namespace turbomind {
 
 struct LlamaAttentionParams {
-    int  rotray_embedding_dim;
-    int  max_position_embeddings;
-    bool use_dynamic_ntk;
-    bool use_logn_attn;
+    int   rotray_embedding_dim;
+    float rotary_emb_base;
```
Comment (lzhangzz) [inline src/turbomind/models/llama/llama_params.h:9]: rotary_embedding_base

File: src/turbomind/triton_backend/llama/LlamaTritonModel.cc [hunk 4]
```diff
@@ -137,6 +137,7 @@ LlamaTritonModel<T>::LlamaTritonModel(size_t      tensor_para_size,
     group_size_            = reader.GetInteger("llama", "group_size", 0);
 
     attn_params_.rotray_embedding_dim    = reader.GetInteger("llama", "rotary_embedding");
+    attn_params_.rotary_emb_base         = reader.GetFloat("llama", "rope_theta", 10000.0f);
```
Comment (lzhangzz) [inline src/turbomind/triton_backend/llama/LlamaTritonModel.cc:140]: rotary_embedding_base

**Comment by michaelroyzen** [comment]
Is this ready to merge? I'm eager to use CodeLlama with LMDeploy.

**Comment by vansin** [comment]
> Is this ready to merge? I'm eager to use CodeLlama with LMDeploy.

Hi, Could you tell me if phind.com is deployed with lmdeploy?

**Comment by lvhan028** [comment]
> Is this ready to merge? I'm eager to use CodeLlama with LMDeploy.

We hope to merge this PR next Monday

**Review by grimoire — APPROVED** [review]

**Review by AllentDan — APPROVED** [review]
LGTM

**Review by lzhangzz — APPROVED** [review]

**Pull request merged**